### DISCUSSION TITLE
Docbook to GFM: When the `<note>` element has no `<title>`, it's not converted to the `note` markdown representation

### DISCUSSION THREAD

**Discussion by tomaskrulis** [jgm/pandoc#9569] [Q&A]
I would like to ask the community about the following behavior:

I have the following docbook XML as an MWE:

```
<?xml version="1.0"?>
<section xmlns="http://docbook.org/ns/docbook" xmlns:xinfo="http://ns.expertinfo.se/cms/xmlns/1.0" xml:id="UUID-ec928802-2645-4d29-3bc6-b59da32feeed" version="5.0" xml:lang="en" xinfo:resource="UUID-ec928802-2645-4d29-3bc6-b59da32feeed" xinfo:resource-id="2201157" xinfo:resource-type="component" xinfo:resource-title="Standard app article template - Jen" xinfo:resource-titlelabel="" xinfo:version-major="1" xinfo:version-minor="0">
  <title>Note test.</title>
      <note>
        <para>A note without title not recognized, regardless of the <code>alerts</code> extension.</para>
      </note>
        <para>Spacer.</para>
    <note>
      <title>
        A note with a title. Works, but the title is discarded.
      </title>
      <para>Some content</para>
    </note>
  </section>
```

The first `<note>` element doesn't have the `<title>`, the second does.

WIth the current pandoc version, when converting with:

```
pandoc -f docbook -t gfm --wrap=none -o mwe.md mwe.xml
```

I get the following conversion result:

```
# Note test.

<div class="note">

A note without title not recognized, regardless of the `alerts` extension.

</div>

Spacer.

> [!NOTE]
> Some content
```

It seems like the `<note>` has to have the `<title>` element to be recognized as a `note`. Even empty `<title>` element makes it work.

Is this intended? Would it be possible to make the `<title>` element within the `<note>` element optional?

I think that the responsible part in the Markdown writer is this: [permalink to the `blockToMarkdown' function`](https://github.com/jgm/pandoc/blob/32280fd919ca2bdba27b5beff9182d8691eb6c1d/src/Text/Pandoc/Writers/Markdown.hs#L375).

Thank you for any guidance on this!

**Comment by jgm** [comment]
Not sure whether it would be better to improve this by (a) having the docbook reader add an empty title Div or by (b) having the markdown writer not care about whether the title Div is present.

(a) is a bit more conservative; it only affects docbook.  (b) might risk capturing things that aren't intended as admonitions? I'm not sure.

Here's a patch that works for (b):
```
diff --git a/src/Text/Pandoc/Writers/Markdown.hs b/src/Text/Pandoc/Writers/Markdown.hs
index 941557e0e..f833d669c 100644
--- a/src/Text/Pandoc/Writers/Markdown.hs
+++ b/src/Text/Pandoc/Writers/Markdown.hs
@@ -372,15 +372,23 @@ blockToMarkdown' opts (Div attrs@(_,classes,_) bs)
   | isEnabled Ext_alerts opts
   , (cls:_) <- classes
   , cls `elem` ["note", "tip", "warning", "caution", "important"]
-  , (Div ("", ["title"], []) _ : Para ils : bs') <- bs
-   = blockToMarkdown' opts $ BlockQuote $
-       (Para (RawInline (Format "markdown") (case cls of
-         "note" -> "[!NOTE]\n"
-         "tip" -> "[!TIP]\n"
-         "warning" -> "[!WARNING]\n"
-         "caution" -> "[!CAUTION]\n"
-         "important" -> "[!IMPORTANT]\n"
-         _ -> "[!NOTE]\n") : ils)) : bs'
+   = do let admonition =
+              case cls of
+                 "note" -> "[!NOTE]"
+                 "tip" -> "[!TIP]"
+                 "warning" -> "[!WARNING]"
+                 "caution" -> "[!CAUTION]"
+                 "important" -> "[!IMPORTANT]"
+                 _ -> "[!NOTE]"
+        let dropTitle (Div ("", ["title"], []) _ : bs') = bs'
+            dropTitle bs' = bs'
+        variant <- asks envVariant
+        let leader
+             | isEnabled Ext_literate_haskell opts = " > "
+             | variant == PlainText = "  "
+             | otherwise            = "> "
+        contents <- blockListToMarkdown opts (dropTitle bs)
+        pure $ prefixed leader (admonition $$ contents) <> blankline
   | otherwise = do
     contents <- blockListToMarkdown opts bs
     variant <- asks envVariant
```

**Reply by tomaskrulis** [reply]
Thank you for the quick reply!

In my opinion, b) makes the most sense in general, since, for markdown, the content of the `<title>` is discarded anyway. But, since markdown is so much used and this change could theoretically break things, it might be better to have a larger discussion forum or audience. Is it possible?

Without a proper discussion, a) would be safer and would work practically just as well. I just think that conceptually, a Writer requiring an element that is discarded in the output anyway is unexpected.

EDIT: I also checked the [docbook specification](https://tdg.docbook.org/tdg/5.2/note) and it doesn't have any other `<note>` elements. So it should be safe from that side.

In the meantime, I tried to create a Lua filter that adds an empty `<title>` element to admonitions that don't use it.

What I got so far:

```lua
local formats = {
  'markdown',
  'gfm',
  'json',  -- testing
  'native' -- testing
}

local admonitions = {
  'note',
  'caution',
  'warning',
  'tip'
}

function Div(elem)
  for _, format in ipairs(formats) do
    if FORMAT:match(format) then
      for _, admonition in ipairs(admonitions) do
        if elem.classes:includes(admonition) then -- OK

          for _, block in ipairs(elem.content) do
            if block.t == 'Div' and block.classes:includes('title') then -- OK
              return nil
            end
          end

          return pandoc.Div({
              pandoc.Div("",
                pandoc.Attr('', {'title'}, {})),  -- OK
--            pandoc.Para (pandoc.Str ("test")) -- testing
              elem.content -- <-- I thought this is a list of Blocks, but apparently, it's a table
            },
            pandoc.Attr('', {'note'}, {}))
        end
      end
    end
  end
end
```

I am struggling with retrieving the original `<div>` content. The content doesn't have to be just one paragraph. In theory, it can be multiple, or there can be lists, or images. So the original content can be multiple blocks, or just one paragraph ...

I also tried with the `insert()` function, but I couldn't figure that out either.

EDIT2: Edited the code for a better version. I thought that `Div.content` is a list of Blocks already, but, apparently, it's a `table`, and I can't figure out how to convert from a `table` to a list of Blocks.

Is it OK if I ask here, or should I create a new discussion? I hope this doesn't switch the theme too much.

**Reply by jgm** [reply]
I think you want something like
```
elem.content:insert([put new block here], 1)
```

**Reply by jgm** [reply]
As for a vs b, we can see if anyone else comments here.

**Reply by tomaskrulis** [reply]
I am not successful with the `insert` function either:

```lua
local formats = {
  'markdown',
  'gfm',
  'json',  -- testing
  'native' -- testing
}

local admonitions = {
  'note',
  'caution',
  'warning',
  'tip'
}

function Div(elem)
  for _, format in ipairs(formats) do
    if FORMAT:match(format) then
      for _, admonition in ipairs(admonitions) do
        if elem.classes:includes(admonition) then -- OK
          for _, block in ipairs(elem.content) do
            if block.t == 'Div' and block.classes:includes('title') then -- OK
              return nil
            else
--              return pandoc.Str("test") -- works, the original Div is replaced
              return elem.content:insert(pandoc.Div("",
                pandoc.Attr('', {'title'}, {}))) -- no insertion happens
            end
          end
        end
      end
    end
  end
end
```

It seems to me like I am making a syntax error or that I am not grasping a bit of the concept of creating new `Div` elements.

I tried also `elem.content:insert(1, pandoc.Div("", pandoc.Attr('', {'title'}, {})))` ; the other way around `elem.content:insert(pandoc.Div("", pandoc.Attr('', {'title'}, {})), 1)` gives me a syntax error.

**Reply by jgm** [reply]
Right, the index has to come first, I always forget the order.

I think what you want is
```lua
elem.content:insert(1, pandoc.Div("", pandoc.Attr('', {'title'}, {})))
return elem
```

**Reply by tomaskrulis** [reply]
This works perfectly, thank you! I tried a few options, didn't try to separate the `insert` method call and the return statement. I thought that the `insert` returns the modified list.

If it's OK, I will wait for follow-up discussion, but if there's none, I would mark the lua filter as a solution, since it achieves the result I need without any potential breaking changes :)

**Reply by jgm** [reply]
I still think we should do either (a) or (b) in pandoc. Neither is difficult.  You could open an issue for this?

**Reply by tomaskrulis** [reply]
Sure thing, I created: https://github.com/jgm/pandoc/issues/9590

Labeled as a feature request, since it doesn't seem like this is a "bug" per se.
